File: hackerrank/problem solving/algorithms/1 - warmup/3 - compare the triplets/with morfo library/compare_the_triplets.cpp
// https://www.hackerrank.com/challenges/compare-the-triplets/problem?isFullScreen=true
// C++23

import std;
import morfo;

using namespace std;

pair<int, int> compute_triplet_scores(const vector<int>& triplet_a, const vector<int>& triplet_b);

int main()
{
    constexpr int n{ 3 };
    vector<int> triplet_a = morfo::read(n);
    vector<int> triplet_b = morfo::read(n);
    auto [score_a, score_b] { compute_triplet_scores(triplet_a, triplet_b) };
    println("{} {}", score_a, score_b);
    return 0;
}

// n: length of a and b = 3
// T: O(3) => O(1)
// S: O(1) extra space
pair<int, int> compute_triplet_scores(const vector<int>& triplet_a, const vector<int>& triplet_b)
{
    int score_a{}, score_b{};

    for (auto [ta, tb] : views::zip(triplet_a, triplet_b))
        if (ta > tb)
            ++score_a;
        else if (tb > ta)
            ++score_b;

    return { score_a, score_b };
}

File: hackerrank/problem solving/algorithms/1 - warmup/3 - compare the triplets/with morfo library/compare_the_triplets.rb
# https://www.hackerrank.com/challenges/compare-the-triplets/problem?isFullScreen=true

require 'morfo'

def main
  triplet_a = Morfo.readln(&:to_i)
  triplet_b = Morfo.readln(&:to_i)
  score_a, score_b = compare_triplets(triplet_a, triplet_b)
  puts "#{score_a} #{score_b}"
end

# n: length of a and b = 3
# T: O(3) => O(1)
# S: O(1) extra space
def compare_triplets(triplet_a, triplet_b)
  score_a, score_b = 0, 0

  triplet_a.zip(triplet_b).each do |ta, tb|
    if ta > tb
      score_a += 1
    elsif tb > ta
      score_b += 1
    end
  end

  [score_a, score_b]
end

main if __FILE__ == $0
